include/mbedtls/fsio.h

```diff
@@ -0,0 +1,246 @@
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *  This file is part of mbed TLS (https://tls.mbed.org)
+ */
+#ifndef MBEDTLS_FILE_H
+#define MBEDTLS_FILE_H
+
+#if !defined(MBEDTLS_CONFIG_FILE)
+#include "config.h"
+#else
+#include MBEDTLS_CONFIG_FILE
+#endif
+
+#include <stddef.h>
+#include <stdint.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#if defined(MBEDTLS_FS_IO)
+/* If MBEDTLS_FS_IO is enabled then file IO functions should be made available
+ * via standard library or platform specific implementation. */
+
+
+/**
+ * Structure returned by mbedtls_stat().
+ *
+ * \note Future versions of the library may add more fields at the end of this
+ * 	 structure.
+ */
+typedef struct mbedtls_stat_tag
+{
+    uint16_t    type;   /** File type */
+/** A file that can be accessed as a stream, i.e. opened with mbedtls_fopen().
+ */
+#define MBEDTLS_FSIO_DT_FILE    0
+/** A directory that can be opened with mbedtls_opendir(). */
+#define MBEDTLS_FSIO_DT_DIR     1
+/** A directory entry that is recognized neither as a stream file nor as a
+ * directory. */
+#define MBEDTLS_FSIO_DT_OTHER   2
+} mbedtls_stat_t;
+
+/* Default and alternative implementation specific interfaces. */
+#if !defined(MBEDTLS_FS_IO_ALT)
+#include <stdio.h>
+#include <dirent.h>
+
+typedef FILE *  mbedtls_file_t;
+#define mbedtls_fread( buf, size, stream )      fread( buf, 1, size, stream )
+#define mbedtls_fwrite( buf, size, stream )     fwrite( buf, 1, size, stream )
+#define mbedtls_fgets       fgets
+#define mbedtls_fclose      fclose
+#define mbedtls_ferror      ferror
+#define mbedtls_fseek       fseek
+#define mbedtls_ftell       ftell
+#define MBEDTLS_SEEK_SET    SEEK_SET
+#define MBEDTLS_SEEK_CUR    SEEK_CUR
+#define MBEDTLS_SEEK_END    SEEK_END
+#define MBEDTLS_FILE_INVALID    NULL
+
+typedef DIR *   mbedtls_dir_t;
+#define mbedtls_opendir         opendir
+#define mbedtls_closedir        closedir
+#define MBEDTLS_DIR_INVALID     NULL
+
+#else /* !MBEDTLS_FS_IO_ALT */
+
+/**
+ * file handle.
+ *
+ */
+typedef int32_t mbedtls_file_t;
+#define MBEDTLS_FILE_INVALID    -1
+
+/**
+ * dir handle.
+ *
+ */
+typedef int32_t mbedtls_dir_t;
+#define MBEDTLS_DIR_INVALID     -1
+
+/**
+ * Definition of reference for seek offset used by mbedtls_fseek().
+ */
+#define MBEDTLS_SEEK_SET    0
+#define MBEDTLS_SEEK_CUR    1
+#define MBEDTLS_SEEK_END    2
+
+/**
+ * \brief          Read file. Follows standard C fread interface.
+ *
+ * \param ptr      Pointer to output buffer
+ * \param size     Size of output buffer.
+ * \param stream   File handle (mbedtls_file_t).
+ *
+ * \return         Number of items read.
+ */
+size_t mbedtls_fread( void *ptr, size_t size, mbedtls_file_t stream );
+
+/**
+ * \brief          Write file. Follows standard C fwrite interface.
+ *
```

Comment:
Here the interface differs from the standard `fwrite` in that granularity argument is missing.

Reply:
Indeed, and likewise `fread`. We only implement the parts of fread/fwrite that people actually use. And the return value needs to be updated from “number of items” to “number of bytes”.